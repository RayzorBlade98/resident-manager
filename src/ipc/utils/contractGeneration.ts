import {
  convertAddressToCityString,
  convertAddressToStreetString,
} from '../../utils/address/address.utils';
import { convertNameToString } from '../../utils/name/name.utils';
import contractTemplate from '_/assets/contract/contractTemplate.md';
import landlordCompanyTemplate from '_/assets/contract/landlordCompanyTemplate.md';
import Landlord from '_/models/landlord/landlord';
import Property from '_/models/property/property';
import { Resident } from '_/models/resident/resident';

export type ContractGenerationArgs = {
  landlord: Landlord;
  resident: Resident;
  property: Property;
};

const placeholderLabels = {
  landlordName: 'LANDLORD_NAME',
  landlordStreet: 'LANDLORD_STREET',
  landlordCity: 'LANDLORD_CITY',
  landlordPhone: 'LANDLORD_PHONE', // Todo
  landlordEmail: 'LANDLORD_EMAIL', // Todo
  landlordCompany: 'LANDLORD_COMPANY',
  residentName: 'RESIDENT_NAME',
  residentStreet: 'RESIDENT_STREET', // Todo
  residentCity: 'RESIDENT_CITY', // Todo
  residentPhone: 'RESIDENT_PHONE', // Todo
  propertyStreet: 'PROPERTY_STREET',
  propertyCity: 'PROPERTY_CITY',
  apartmentFloor: 'APARTMENT_FLOOR',
  apartmentLocation: 'APARTMENT_LOCATION',
  apartmentRoomsGeneric: 'APARTMENT_ROOMS_GENERIC',
  apartmentRoomsKitchen: 'APARTMENT_ROOMS_KITCHEN',
  apartmentRoomsBath: 'APARTMENT_ROOMS_BATH',
  apartmentRoomsHallway: 'APARTMENT_ROOMS_HALLWAY',
  apartmentRoomsBasement: 'APARTMENT_ROOMS_BASEMENT',
  apartmentRoomsGarden: 'APARTMENT_ROOMS_GARDEN', // Todo
  parkingSpaceCount: 'PARKING_SPACE_COUNT',
  keysApartment: 'KEYS_APARTMENT',
  keysBasement: 'KEYS_BASEMENT',
  keysAttic: 'KEYS_ATTIC',
  keysFrontdoor: 'KEYS_FRONTDOOR',
  keysMailbox: 'KEYS_MAILBOX',
} satisfies Record<string, string>;

const blockPlaceholderLabels = {
  landlordCompany: 'LANDLORD_COMPANY_BLOCK',
} satisfies Record<string, string>;

/**
 * Generates the contract as markdown string
 * @returns markdown string of the generated contract
 */
export function generateContractMarkdown(args: ContractGenerationArgs): string {
  const apartment = args.property.apartments.find(
    (a) => a.id === args.resident.apartmentId,
  )!;

  const placeholders = {
    [placeholderLabels.landlordName]: convertNameToString(
      args.landlord.representative,
    ),
    [placeholderLabels.landlordStreet]: convertAddressToStreetString(
      args.landlord.address,
    ),
    [placeholderLabels.landlordCity]: convertAddressToCityString(
      args.landlord.address,
    ),
    [placeholderLabels.landlordCompany]: args.landlord.company ?? '',
    [placeholderLabels.residentName]: convertNameToString(args.resident.name),
    [placeholderLabels.propertyStreet]: convertAddressToStreetString(
      args.property.address,
    ),
    [placeholderLabels.propertyCity]: convertAddressToCityString(
      args.property.address,
    ),
    [placeholderLabels.apartmentFloor]: apartment.floor,
    [placeholderLabels.apartmentLocation]: apartment.location,
    [placeholderLabels.apartmentRoomsGeneric]:
      apartment.rooms.generic.toString(),
    [placeholderLabels.apartmentRoomsKitchen]:
      apartment.rooms.kitchen.toString(),
    [placeholderLabels.apartmentRoomsBath]: apartment.rooms.bath.toString(),
    [placeholderLabels.apartmentRoomsBasement]:
      apartment.rooms.basement.toString(),
    [placeholderLabels.apartmentRoomsHallway]:
      apartment.rooms.hallway.toString(),
    [placeholderLabels.parkingSpaceCount]: args.resident.parkingSpaceId ? '1' : '0',
    [placeholderLabels.keysApartment]: args.resident.keys.apartment.toString(),
    [placeholderLabels.keysBasement]: args.resident.keys.basement.toString(),
    [placeholderLabels.keysAttic]: args.resident.keys.attic.toString(),
    [placeholderLabels.keysFrontdoor]: args.resident.keys.frontDoor.toString(),
    [placeholderLabels.keysMailbox]: args.resident.keys.mailbox.toString(),
  };

  let contract = contractTemplate;

  // Landlord company
  contract = contract.replace(
    getPlaceholder(blockPlaceholderLabels.landlordCompany),
    args.landlord.company ? landlordCompanyTemplate : '',
  );

  return replaceAllPlaceholders(contract, placeholders);
}

function replaceAllPlaceholders(
  contract: string,
  replacements: Record<string, string>,
): string {
  Object.entries(replacements).forEach(([label, replacement]) => {
    contract = contract.replace(getPlaceholder(label), replacement);
  });
  return contract;
}

function getPlaceholder(label: string): string {
  return `{{${label}}}`;
}
