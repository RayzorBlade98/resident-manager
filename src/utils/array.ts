/**
 * This module provides functionality regarding `Array` objects
 * like creating the cartesian product of multiple arrays.
 */

/**
 * Creates an array containing a sequence of numbers for a given range.
 *
 * Examples:
 * - `range(0, 4) := [0, 1, 2, 3, 4]`
 * - `range (0, 4, 2) := [0, 2, 4]`
 * - `range (0, 5, 2) := [0, 2, 4]`
 *
 * @param start start value of the sequence
 * @param stop end value of the sequence
 * @param step space between each element of the sequence (default: 1)
 * @returns sequence of numbers for the given range
 */
export function range(start: number, stop: number, step: number = 1): number[] {
  return [...Array(Math.floor((stop - start) / step) + 1)].map(
    (_, i) => start + i * step,
  );
}

/**
 * Creates the cartesian product for any number of given array.
 *
 * Examples:
 * - `cartesianProduct([1, 2], [3, 4]) := [[1, 3], [1, 4], [2, 3], [2, 4]]`
 * - `cartesianProduct([1, 2], [3, 4], [5, 6]) := [[1, 3, 5], [1, 3, 6], [1, 4, 5], [1, 4, 6], [2, 3, 5], [2, 3, 6], [2, 4, 5], [2, 4, 6]]`
 *
 * @param allEntries any number of array that should be combined
 * @returns array containing the cartesian product over all given arrays
 */
export function cartesianProduct<T>(...allEntries: T[][]): T[][] {
  return allEntries.reduce<T[][]>(
    (results, entries) =>
      results
        .map((result) => entries.map((entry) => result.concat([entry])))
        .reduce((subResults, result) => subResults.concat(result), []),
    [[]],
  );
}
