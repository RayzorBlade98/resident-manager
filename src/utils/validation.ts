/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-use-before-define */

import _ from 'lodash';

/**
 * Object containing an error message for each key of `T` that had an invalid value or `undefined` if it had a valid value.
 */
export type ValidationErrorMessages<T> = {
  [k in keyof T]?: string;
};

/**
 * Enum containing all available validation errors.
 */
export enum ValidationConstraint {
  /**
   * ValidationConstraint that checks if the tested string is not empty.
   */
  NoEmptyString,

  /**
   * ValidationConstraint that checks if the tested value is between `1` and `12`.
   */
  Month,

  /**
   * ValidationConstraint that checks if the tested number is a valid currency value.
   */
  Currency,
}

/**
 * Creates a function that validates an object of type `T`.
 * @param errors `ValidationErrors` that should be tested for each key of `T`
 * @returns validation function
 */
export function createValidationFunction<T>(errors: {
  [k in keyof T]?: ValidationConstraint[];
}): (
    object: T,
    key?: keyof T | undefined,
  ) => ValidationErrorMessages<T> | string | undefined {
  /**
   * Validates each key of an object of type `T`
   * @param object object of type `T` that should be validated
   * @param key optional key of `T`. If provided only the error message for this key is returned
   * @returns error message for each invalid key of `object`
   */
  function validationFunction(
    object: T,
    key: keyof T | undefined = undefined,
  ): ValidationErrorMessages<T> | string | undefined {
    const errorMessages: Partial<ValidationErrorMessages<T>> = {};
    // eslint-disable-next-line no-restricted-syntax
    for (const [k, value] of Object.entries(errors)) {
      errorMessages[k as keyof T] = undefined;
      if (value === undefined) continue;
      // eslint-disable-next-line no-restricted-syntax
      for (const error of value as ValidationConstraint[]) {
        const errorMessage = CONSTRAINT_FUNCTIONS[error](object[k as keyof T]);
        if (errorMessage) {
          errorMessages[k as keyof T] = errorMessage;
          break;
        }
      }
    }
    if (key) {
      return errorMessages[key];
    }
    return errorMessages;
  }
  return validationFunction;
}

/**
 * The contraints that each key of `T` must fulfill
 */
type ValidationConstraints<T> = {
  [k in keyof T]?: ValidationConstraint;
};

/**
 * Class that provide functionalities to validate objects of type `T`
 */
export class Validator<T extends object> {
  private constraints: ValidationConstraints<T>;

  /**
   * @param constraints constraints that the object values need to fulfill
   */
  public constructor(constraints: ValidationConstraints<T>) {
    this.constraints = constraints;
  }

  /**
   * Validates the given object
   * @param toValidate object that should be validated
   * @param invalidOnly whether only the invalid keys should be included into the returned object or not
   * @returns object containing the error messages for invalid keys of the validated object and undefined for the valid keys.
   */
  public validate(
    toValidate: Partial<T>,
    invalidOnly = true,
  ): ValidationErrorMessages<T> {
    const errorMessages: ValidationErrorMessages<T> = {};
    Object.entries(this.constraints).forEach(([_key, _constraint]) => {
      const key = _key as unknown as keyof T;
      const constraint = _constraint as ValidationConstraint;
      if (constraint === undefined) return;
      errorMessages[key] = CONSTRAINT_FUNCTIONS[constraint](toValidate[key]);
    });

    if (!invalidOnly) {
      return errorMessages;
    }

    return _.pickBy(
      errorMessages,
      (v, _k) => v !== undefined,
    ) as ValidationErrorMessages<T>;
  }

  /**
   * Validates the new values that are different from the old values
   * @param oldValues old values that the new values are compared to
   * @param newValues new values that should be validated
   * @returns object containing the error messages of all keys that have different values than. Only invalid keys will be included.
   */
  public validateDifference(
    oldValues: T,
    newValues: T,
  ): ValidationErrorMessages<T> {
    const difference = _.pickBy(
      newValues,
      (v, k) => oldValues[k as keyof T] !== v,
    );
    return this.validate(difference, false);
  }
}

/**
 * Object containing a function that checks if a given values is invalid for each `ValidationError`
 */
const CONSTRAINT_FUNCTIONS: {
  [k in ValidationConstraint]: (value: any) => string | undefined;
} = {
  [ValidationConstraint.Month]: monthConstraint,
  [ValidationConstraint.NoEmptyString]: noEmptyStringConstraint,
  [ValidationConstraint.Currency]: currencyConstraint,
};

/**
 * Object containing all error messages
 */
const ERROR_MESSAGES = {
  EMPTY: 'Darf nicht leer sein!',
  LT_ZERO: 'Muss größer als 0 sein!',
  NO_INTEGER: 'Muss eine ganze Zahl sein!',
  NO_MONTH: 'Muss zwischen 1 und 12 sein!',
  NO_NUMBER: 'Muss eine Zahl sein!',
};

/**
 * GENERIC CONSTRAINT FUNCTIONS
 */

/**
 * Checks if the value is a number
 */
function isNumberConstraint(
  value: string | number | undefined,
): string | undefined {
  if (!value || value === '') {
    return ERROR_MESSAGES.EMPTY;
  }
  if (Number.isNaN(Number(value))) {
    return ERROR_MESSAGES.NO_NUMBER;
  }
  return undefined;
}

/**
 * Checks if the value is an integer
 */
function isIntegerConstraint(
  value: string | number | undefined,
): string | undefined {
  if (isNumberConstraint(value) || Number(value) % 1 !== 0) {
    return ERROR_MESSAGES.NO_INTEGER;
  }
  return undefined;
}

/**
 * SPECIFIC CONSTRAINT FUNCTIONS
 */

/**
 * Constraint function for the `Currency` constraint
 */
function currencyConstraint(value: number | undefined): string | undefined {
  if (!value) {
    return ERROR_MESSAGES.EMPTY;
  }
  if (value <= 0) {
    return ERROR_MESSAGES.LT_ZERO;
  }
  return undefined;
}

/**
 * Constraint function for the `NoEmptyString` constraint
 */
function noEmptyStringConstraint(value: string): string | undefined {
  if (value === '') {
    return ERROR_MESSAGES.EMPTY;
  }
  return undefined;
}

/**
 * Constraint function for the `Month` constraint
 */
function monthConstraint(
  value: string | number | undefined,
): string | undefined {
  const isNoInteger = isIntegerConstraint(value);
  if (isNoInteger) {
    return isNoInteger;
  }
  const number = Number(value);
  if (number < 1 || number > 12) {
    return ERROR_MESSAGES.NO_MONTH;
  }
  return undefined;
}
