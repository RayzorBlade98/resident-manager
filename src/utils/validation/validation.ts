/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-use-before-define */

import { RecoilState, selector } from 'recoil';
import Validator from './validator';

/**
 * Object containing an error message for each key of `T` that had an invalid value or `undefined` if it had a valid value.
 */
export type ValidationErrorMessages<T> = {
  [k in keyof T]?: string;
};

/**
 * State that contains all information about a form validation
 */
export type FormValidationState<T extends object> = {
  /**
   * Current form input
   */
  formInput: T;

  /**
   * Current form error messages
   */
  formErrors: ValidationErrorMessages<T>;

  /**
   * Validator that is handling the form validation
   */
  formValidator: Validator<T>;
};

/**
 * State that is using form validation
 */
export type CompleteFormValidationState<
  OtherState,
  InputType extends object,
> = OtherState & {
  formValidation: FormValidationState<InputType>;
};

/**
 * Creates a recoil selector for the form validation state
 * @param recoilState State that contains the validation state
 * @returns Selector for the form validation state
 */
export function createFormValidationStateSelector<
  OtherState,
  InputType extends object,
>(
  recoilState: RecoilState<CompleteFormValidationState<OtherState, InputType>>,
): RecoilState<FormValidationState<InputType>> {
  return selector<FormValidationState<InputType>>({
    key: `${recoilState.key}-formValidation`,
    get: ({ get }) => get(recoilState).formValidation,
    set: ({ set }, newState) => set(recoilState, (state) => {
      const formValidation = newState as FormValidationState<InputType>;
      return {
        ...state,
        formValidation: {
          ...formValidation,
          formErrors: {
            ...formValidation.formErrors,
            ...formValidation.formValidator.validateDifference(
              state.formValidation.formInput,
              formValidation.formInput,
            ),
          },
        },
      };
    }),
  });
}
