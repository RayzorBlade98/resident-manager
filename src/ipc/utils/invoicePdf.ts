import jsPDF from 'jspdf';
import autoTable from 'jspdf-autotable';
import {
  convertAddressToCityString,
  convertAddressToStreetString,
} from '../../utils/address/address.utils';
import { convertCurrencyCentsToString } from '../../utils/currency/currency.utils';
import { convertNameToString } from '../../utils/name/name.utils';
import MonthYear from '_/extensions/date/month_year.extension';
import Invoice from '_/models/invoice/invoice';
import ResidentInvoiceInformation from '_/models/invoice/resident.invoice';

const labels = {
  addresses: {
    residentTitle: 'Mieter:',
  },
  ongoingIncidentals: {
    title: 'Laufende Nebenkosten',
    tableHeader: [
      'Nebenkosten',
      'Abrechnungsart',
      'Gesamtkosten',
      'Mieteranteil',
    ],
  },
  oneTimeIncidentals: {
    title: 'Einmalige Nebenkosten',
    tableHeader: [
      'Nebenkosten',
      'Abrechnungsart',
      'Gesamtkosten',
      'Mieteranteil',
    ],
  },
  waterCosts: {
    title: 'Wasserkosten',
    tableHeader: [
      'Kostenart',
      'Alter Zählerstand',
      'Neuer Zählerstand',
      'Wasserverbrauch',
      'Kosten pro m³',
      'Gesamtkosten',
    ],
    waterUsageCostLabel: 'Wasser',
    sewageCostLabel: 'Abwasser',
  },
  rentPayments: {
    title: 'Mietzahlungen',
    tableHeader: [
      'Monat',
      'Miete',
      'Nebenkosten',
      'Gezahlte Miete',
      'Fehlende Miete',
    ],
  },
};

/**
 * Class that provides functionality to create a pdf invoice for a resident
 */
class InvoicePdfGenerator {
  private readonly invoice: Invoice;

  private readonly residentInformation: ResidentInvoiceInformation;

  private readonly pdf: jsPDF;

  /**
   * @param invoice invoice for which the pdf should be generated
   * @param residentId id of the resident for which the pdf should be generated
   */
  public constructor(invoice: Invoice, residentId: string) {
    this.invoice = invoice;
    this.residentInformation = invoice.residentInformation[residentId];
    this.pdf = new jsPDF();
  }

  /**
   * Generates a `jsPDF` for the provided invoice and resident
   * @returns generated pdf
   */
  public generatePdf(): jsPDF {
    this.addAddresses();

    if (
      Object.keys(this.residentInformation.ongoingIncidentalsCosts).length > 0
    ) {
      this.addOngoingIncidentals();
    }

    if (
      Object.keys(this.residentInformation.oneTimeIncidentalsCosts).length > 0
    ) {
      this.addOneTimeIncidentals();
    }

    if (this.residentInformation.waterCosts.waterUsage > 0) {
      this.addWaterCosts();
    }

    if (this.residentInformation.rentPayments.length > 0) {
      this.addRentPayments();
    }

    return this.pdf;
  }

  private addAddresses(): void {
    const residentAddress = [
      labels.addresses.residentTitle,
      convertNameToString(this.residentInformation.name),
      convertAddressToStreetString(this.invoice.property.address),
      convertAddressToCityString(this.invoice.property.address),
    ].join('\n');

    autoTable(this.pdf, {
      body: [
        [
          {
            content: residentAddress,
            styles: {
              halign: 'left',
            },
          },
          // Todo: Add land lord address
          {
            content: '',
            styles: {
              halign: 'right',
            },
          },
        ],
      ],
      theme: 'plain',
    });
  }

  private addOngoingIncidentals() {
    this.addHeadline(labels.ongoingIncidentals.title);
    this.addTable(
      labels.ongoingIncidentals.tableHeader,
      Object.entries(this.residentInformation.ongoingIncidentalsCosts).map(
        ([incidentals, costs]) => {
          const incidentalsInformation = this.invoice.ongoingIncidentalsInformation[incidentals];
          return [
            incidentalsInformation.name,
            incidentalsInformation.deductionType,
            convertCurrencyCentsToString(incidentalsInformation.totalCost),
            convertCurrencyCentsToString(costs),
          ];
        },
      ),
    );
  }

  private addOneTimeIncidentals() {
    this.addHeadline(labels.oneTimeIncidentals.title);
    this.addTable(
      labels.oneTimeIncidentals.tableHeader,
      Object.entries(this.residentInformation.oneTimeIncidentalsCosts).map(
        ([incidentals, costs]) => {
          const incidentalsInformation = this.invoice.oneTimeIncidentalsInformation[incidentals];
          return [
            incidentalsInformation.name,
            incidentalsInformation.deductionType,
            convertCurrencyCentsToString(incidentalsInformation.totalCost),
            convertCurrencyCentsToString(costs),
          ];
        },
      ),
    );
  }

  private addWaterCosts() {
    const waterCosts = this.residentInformation.waterCosts;

    const waterUsage = [
      `${waterCosts.lastWaterMeterCount} m³`,
      `${waterCosts.currentWaterMeterCount} m³`,
      `${waterCosts.waterUsage} m³`,
    ];

    this.addHeadline(labels.waterCosts.title);
    this.addTable(labels.waterCosts.tableHeader, [
      [
        labels.waterCosts.waterUsageCostLabel,
        ...waterUsage,
        convertCurrencyCentsToString(
          this.invoice.waterCosts.waterUsageCostPerCubicMeter,
        ),
        convertCurrencyCentsToString(waterCosts.waterUsageCosts),
      ],
      [
        labels.waterCosts.sewageCostLabel,
        ...waterUsage,
        convertCurrencyCentsToString(
          this.invoice.waterCosts.sewageCostPerCubicMeter,
        ),
        convertCurrencyCentsToString(waterCosts.sewageCosts),
      ],
    ]);
  }

  private addRentPayments() {
    this.addHeadline(labels.rentPayments.title);
    this.addTable(
      labels.rentPayments.tableHeader,
      this.residentInformation.rentPayments.map((r) => [
        MonthYear.fromDate(r.dueDate).toString(),
        convertCurrencyCentsToString(r.rent),
        convertCurrencyCentsToString(r.incidentals),
        convertCurrencyCentsToString(r.paymentAmount),
        convertCurrencyCentsToString(r.paymentMissing),
      ]),
    );
  }

  private addHeadline(title: string) {
    autoTable(this.pdf, {
      body: [
        [
          {
            content: title,
            styles: {
              halign: 'left',
              fontSize: 14,
            },
          },
        ],
      ],
      theme: 'plain',
    });
  }

  private addTable(head: string[], body: string[][]) {
    autoTable(this.pdf, {
      head: [head],
      body,
      theme: 'striped',
      headStyles: {
        fillColor: '#343a40',
      },
    });
  }
}

export default InvoicePdfGenerator;
