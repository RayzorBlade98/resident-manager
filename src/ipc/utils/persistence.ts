import fs from 'fs';
import path from 'path';
// eslint-disable-next-line import/no-extraneous-dependencies
import { app, dialog } from 'electron';
import jsPDF from 'jspdf';
import mainWindow from '../../electron/windows';
import { prod } from '_/utils/node-env';

/**
 * Returns the directory to which the exported files are saved to
 */
/* istanbul ignore next */
const saveDirectory = () => (prod ? app.getPath('userData') : __dirname);

/**
 * Converts the given object to json and writes it to the specified file
 * @param toExport object that should be exported
 * @param filename file to which the object is saved
 */
export function exportObject<T>(toExport: T, filename: string): void {
  const outputPath = path.join(saveDirectory(), filename);
  const json = JSON.stringify(toExport, null, 4);
  fs.writeFileSync(outputPath, json);
}

/**
 * Imports the specified json file and parses it to an object of type T
 * @param filename file that should be imported
 * @returns `null` if the import failed, otherwise the imported object
 */
export function importObject<T>(filename: string): T | null {
  const inputPath = path.join(saveDirectory(), filename);
  if (!fs.existsSync(inputPath)) {
    return null;
  }
  const json = fs.readFileSync(inputPath).toString();
  return JSON.parse(json) as T;
}

/**
 * Exports a jsPDF object to a file
 * @param pdf jsPDF that should be exported
 * @param filePath file to which the pdf should be exported
 */
export function exportJsPdf(pdf: jsPDF, filePath: string): void {
  const buffer = Buffer.from(pdf.output('arraybuffer'));
  fs.writeFileSync(filePath, buffer);
}

/**
 * Opens a file dialog that lets the user pick a single directory
 * @returns path to the directory if the user picked on, otherwise undefined
 */
export function openDirectoryDialog(): string | undefined {
  return dialog
    .showOpenDialogSync(mainWindow, {
      properties: ['openDirectory'],
    })
    ?.at(0);
}

/**
 * Opens a file dialog that lets the user pick a single file
 * @returns path to the file if the user picked on, otherwise undefined
 */
export function openFileDialog(): string | undefined {
  return dialog
    .showOpenDialogSync(mainWindow, {
      properties: ['openFile'],
      filters: [
        {
          name: 'Pdf',
          extensions: ['pdf'],
        },
      ],
    })
    ?.at(0);
}
