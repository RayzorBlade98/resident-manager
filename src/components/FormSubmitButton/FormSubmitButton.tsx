import { Button } from '@mui/material';
import React from 'react';
import { ValidationErrorMessages, Validator } from '_/utils/validation';

interface FormSubmitButtonProps<T extends object> {
  /**
   * Text of the button
   */
  buttonText: string;

  /**
   * Form inputs that should be validated
   */
  formInput: T;

  /**
   * Validator that validates the form input
   */
  validator: Validator<T>;

  /**
   * Callback that is called when the button gets pressed for a valid input
   */
  onSuccess: () => void;

  /**
   * Callback that is called when the button gets pressed for an invalid input
   * @param errors errormessages of the form input
   */
  onError: (errors: ValidationErrorMessages<T>) => void;
}

/**
 * Button that handles form validation when clicked
 */
function FormSubmitButton<T extends object>(
  props: FormSubmitButtonProps<T>,
): JSX.Element {
  const errors = props.validator.validate(props.formInput);
  const isValidInput = Object.keys(errors).length === 0;

  const onSave = (): void => {
    if (isValidInput) {
      props.onSuccess();
    } else {
      props.onError(errors);
    }
  };

  return (
    <Button
      variant="contained"
      color={isValidInput ? 'primary' : 'error'}
      onClick={onSave}
    >
      {props.buttonText}
    </Button>
  );
}

export default FormSubmitButton;
