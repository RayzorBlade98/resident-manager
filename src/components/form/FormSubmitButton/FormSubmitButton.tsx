import { Button } from '@mui/material';
import React from 'react';
import { RecoilState, useRecoilState } from 'recoil';
import { FormValidationState } from '_/utils/validation/validation';

interface FormSubmitButtonProps<T extends object> {
  /**
   * Text of the button
   */
  buttonText: string;

  /**
   *
   */
  formState: RecoilState<FormValidationState<T>>;

  /**
   * Callback that is called when the button gets pressed for a valid input
   */
  onSuccess: () => void;
}

/**
 * Button that handles form validation when clicked
 */
function FormSubmitButton<T extends object>(
  props: FormSubmitButtonProps<T>,
): JSX.Element {
  const [formState, setFormState] = useRecoilState(props.formState);
  const errors = formState.formValidator.validate(formState.formInput);
  const isValidInput = Object.keys(errors).length === 0;

  const onSave = (): void => {
    if (isValidInput) {
      props.onSuccess();
    } else {
      setFormState((state) => ({ ...state, formErrors: errors }));
    }
  };

  return (
    <Button
      variant="contained"
      color={isValidInput ? 'primary' : 'error'}
      onClick={onSave}
    >
      {props.buttonText}
    </Button>
  );
}

export default FormSubmitButton;
