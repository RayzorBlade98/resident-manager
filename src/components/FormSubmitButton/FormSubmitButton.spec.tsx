/* eslint-disable max-len */

import { RenderResult, fireEvent, render } from '@testing-library/react';
import React from 'react';
import FormSubmitButton from './FormSubmitButton';
import { Validator } from '_/utils/validation';

describe('FormSubmitButton', () => {
  type Test = { success: boolean };
  const validator = new Validator<Test>({});
  const onSuccessMock = jest.fn();
  const onErrorMock = jest.fn();
  let validateSpy: jest.SpyInstance;

  function renderComponent(success: boolean): RenderResult {
    return render(
      <FormSubmitButton<Test>
        buttonText="Testbutton"
        formInput={{ success }}
        validator={validator}
        onSuccess={onSuccessMock}
        onError={onErrorMock}
      />,
    );
  }

  beforeEach(() => {
    validateSpy = jest
      .spyOn(validator, 'validate')
      .mockImplementation((toValidate: Partial<Test>) => (toValidate.success ? {} : { success: 'fail' }));
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('should call onSuccess for valid input', () => {
    // Arrange
    const renderResult = renderComponent(true);

    // Act
    const button = renderResult.getByRole('button');
    fireEvent.click(button);

    // Assert
    expect(validateSpy).toHaveBeenCalledTimes(1);
    expect(onSuccessMock).toHaveBeenCalledTimes(1);
    expect(onErrorMock).toHaveBeenCalledTimes(0);
  });

  test('should call onError for invalid input', () => {
    // Arrange
    const renderResult = renderComponent(false);

    // Act
    const button = renderResult.getByRole('button');
    fireEvent.click(button);

    // Assert
    expect(validateSpy).toHaveBeenCalledTimes(1);
    expect(onSuccessMock).toHaveBeenCalledTimes(0);
    expect(onErrorMock).toHaveBeenCalledTimes(1);
    expect(onErrorMock).toHaveBeenCalledWith({ success: 'fail' });
  });

  test('should match snapshot (valid input)', () => {
    // Arrange
    const renderResult = renderComponent(true);

    // Assert
    expect(renderResult.container).toMatchSnapshot();
  });

  test('should match snapshot (invalid input)', () => {
    // Arrange
    const renderResult = renderComponent(false);

    // Assert
    expect(renderResult.container).toMatchSnapshot();
  });
});
