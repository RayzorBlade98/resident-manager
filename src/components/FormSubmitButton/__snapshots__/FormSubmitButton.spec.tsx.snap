// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSubmitButton should match snapshot (invalid input) 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium css-1sszw6i-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Testbutton
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`FormSubmitButton should match snapshot (valid input) 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Testbutton
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
